File: tools/riskctl.py
#!/usr/bin/env python3
import sys
import yaml
from pathlib import Path

MODEL_DIR = Path("model")
QUESTIONS_DIR = MODEL_DIR / "questions"
RULES_DIR = MODEL_DIR / "rules"
CONTROLS_DIR = MODEL_DIR / "controls"

BASE_QUESTIONS_FILE = QUESTIONS_DIR / "base.questions.yaml"
TRIGGERS_FILE = RULES_DIR / "triggers.rules.yaml"
CONTROLS_RULES_FILE = RULES_DIR / "controls.rules.yaml"
CONTROLS_CATALOG_FILE = CONTROLS_DIR / "controls.catalog.yaml"

ALLOWED_ACTIVATION_PHASES = {"design", "pre_go_live", "runtime", "post_go_live"}
ALLOWED_EVIDENCE_TYPES = {"config", "pipeline", "log", "document"}


def deep_get(d, path: str):
    cur = d
    for p in path.split("."):
        if not isinstance(cur, dict) or p not in cur:
            return None
        cur = cur[p]
    return cur


def matches_condition(facts: dict, cond: dict) -> bool:
    """
    Condition keys can be:
      - "base_key"  (shorthand for base.base_key)
      - "namespace.key" (full path)
    Values can be scalars or membership checks when actual is a list.
    """
    for k, expected in cond.items():
        if "." in k:
            actual = deep_get(facts, k)
        else:
            actual = deep_get(facts, f"base.{k}")

        # membership check: actual list contains expected scalar
        if isinstance(actual, list) and not isinstance(expected, list):
            if expected not in actual:
                return False
        else:
            if actual != expected:
                return False
    return True


def load_yaml(path: Path) -> dict:
    return yaml.safe_load(path.read_text(encoding="utf-8"))


def warn(msg: str):
    print(f"WARN: {msg}", file=sys.stderr)


def validate_controls_catalog(catalog_doc: dict):
    controls = catalog_doc.get("controls", [])
    if not isinstance(controls, list):
        warn("controls catalog: expected 'controls' to be a list")
        return

    for c in controls:
        if not isinstance(c, dict):
            warn("controls catalog: found non-object control entry")
            continue

        cid = c.get("id", "(missing id)")

        phase = c.get("activation_phase")
        if phase is not None and phase not in ALLOWED_ACTIVATION_PHASES:
            warn(
                f"controls catalog: {cid} has unknown activation_phase '{phase}' (allowed: {sorted(ALLOWED_ACTIVATION_PHASES)})"
            )

        evidence = c.get("evidence_type", [])
        if isinstance(evidence, str):
            evidence = [evidence]
        if not isinstance(evidence, list):
            warn(f"controls catalog: {cid} evidence_type should be a list")
            continue

        for e in evidence:
            if e not in ALLOWED_EVIDENCE_TYPES:
                warn(
                    f"controls catalog: {cid} has unknown evidence_type '{e}' (allowed: {sorted(ALLOWED_EVIDENCE_TYPES)})"
                )


def validate_controls_rules(rules_doc: dict, catalog: dict[str, dict]):
    rules = rules_doc.get("rules", [])
    if not isinstance(rules, list):
        warn("controls rules: expected 'rules' to be a list")
        return

    for idx, rule in enumerate(rules, start=1):
        if not isinstance(rule, dict):
            warn(f"controls rules: rule #{idx} is not an object")
            continue

        required = rule.get("require", [])
        if isinstance(required, str):
            required = [required]

        if not isinstance(required, list):
            warn(f"controls rules: rule #{idx} has non-list require")
            continue

        for cid in required:
            if cid not in catalog:
                cond = rule.get("when", {})
                warn(f"controls rules: rule #{idx} references missing control '{cid}' (when: {cond})")


def collect_base_question_ids() -> list[str]:
    doc = load_yaml(BASE_QUESTIONS_FILE)
    return [q["id"] for q in doc.get("questions", [])]


def derive_activated_domains(facts: dict) -> list[str]:
    doc = load_yaml(TRIGGERS_FILE)
    activated = set()

    for t in doc.get("triggers", []):
        cond = t.get("when", {})
        if matches_condition(facts, cond):
            for domain in t.get("activate", []):
                activated.add(domain)

    # stable output order
    return sorted(activated)


def domain_questions_file(domain: str) -> Path:
    return QUESTIONS_DIR / f"{domain}.questions.yaml"


def list_domain_questions(domain: str) -> list[dict]:
    path = domain_questions_file(domain)
    if not path.exists():
        return []
    doc = load_yaml(path)
    return doc.get("questions", [])


def derive_controls(facts: dict) -> tuple[dict, dict]:
    """
    Returns:
      derived: control_id -> list[cond] (why)
      catalog: control_id -> control metadata
    """
    rules_doc = load_yaml(CONTROLS_RULES_FILE)
    catalog_doc = load_yaml(CONTROLS_CATALOG_FILE)

    validate_controls_catalog(catalog_doc)

    catalog = {c["id"]: c for c in catalog_doc.get("controls", [])}
    validate_controls_rules(rules_doc, catalog)
    derived: dict[str, list[dict]] = {}

    for rule in rules_doc.get("rules", []):
        cond = rule.get("when", {})
        if matches_condition(facts, cond):
            for cid in rule.get("require", []):
                derived.setdefault(cid, []).append(cond)

    return derived, catalog


def print_required_questions(base_ids: list[str], activated_domains: list[str], facts: dict):
    print("Required questions (progressive disclosure):")
    print("- base:")
    for qid in base_ids:
        val = deep_get(facts, f"base.{qid}")
        marker = "answered" if val is not None else "missing"
        print(f"  - {qid} ({marker})")

    for domain in activated_domains:
        print(f"- {domain}:")
        questions = list_domain_questions(domain)
        if not questions:
            print("  - (no question file found)")
            continue

        for q in questions:
            qid = q["id"]
            val = deep_get(facts, f"{domain}.{qid}")
            marker = "answered" if val is not None else "missing"
            print(f"  - {domain}.{qid} ({marker})")

    print()


def print_activated_domains(activated_domains: list[str]):
    print("Activated domains:")
    if not activated_domains:
        print("- (none)")
    else:
        for d in activated_domains:
            print(f"- {d} ({domain_questions_file(d)})")
    print()


def print_controls(derived: dict, catalog: dict):
    print("Derived controls:")
    if not derived:
        print("- (none)")
        return

    for cid in sorted(derived.keys()):
        c = catalog.get(cid, {"id": cid, "title": "(missing from catalog)"})
        title = c.get("title", "(no title)")
        scope = c.get("scope", "unknown")
        intent = c.get("enforcement_intent", "unknown")
        phase = c.get("activation_phase", "unknown")
        evidence = c.get("evidence_type", [])

        print(f"- {cid}: {title}")
        print(f"  scope: {scope}")
        print(f"  enforcement_intent: {intent}")
        print(f"  activation_phase: {phase}")
        if evidence:
            print(f"  evidence_type: {evidence}")
        print("  because:")
        for why in derived[cid]:
            print(f"    - when: {why}")
    print()


def main():
    if len(sys.argv) < 3 or sys.argv[1] != "evaluate":
        print("Usage: riskctl.py evaluate <facts.yaml>")
        sys.exit(2)

    facts_path = Path(sys.argv[2])
    if not facts_path.exists():
        print(f"Facts file not found: {facts_path}")
        sys.exit(2)

    facts = load_yaml(facts_path)

    # 1) Base + triggers => activated domains
    base_ids = collect_base_question_ids()
    activated_domains = derive_activated_domains(facts)

    print_activated_domains(activated_domains)
    print_required_questions(base_ids, activated_domains, facts)

    # 2) Controls
    derived, catalog = derive_controls(facts)
    print_controls(derived, catalog)


if __name__ == "__main__":
    main()

File: tools/riskctl.ts
#!/usr/bin/env bun
/**
 * Risk Assistant CLI
 *
 * Usage:
 *   bun tools/riskctl.ts evaluate <facts.yaml> [--model-dir model]
 *   bun tools/riskctl.ts diff <facts.yaml> --old model-v1 --new model-v2
 */

import { parseArgs } from "node:util";
import { existsSync } from "node:fs";
import { resolve } from "node:path";
import { evaluateFacts } from "../web/src/lib/evaluator";
import { modelPaths, getModelVersion } from "../web/src/lib/model";
import { loadYamlFile } from "../web/src/lib/yaml";
import type { Facts, EvaluateResult } from "../web/src/lib/evaluator";

const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const DIM = "\x1b[2m";
const CYAN = "\x1b[36m";
const GREEN = "\x1b[32m";
const YELLOW = "\x1b[33m";
const RED = "\x1b[31m";
const BLUE = "\x1b[34m";

function findRepoRoot(startPath: string): string {
	let current = startPath;
	while (current !== "/") {
		if (existsSync(resolve(current, "model"))) {
			return current;
		}
		const parent = resolve(current, "..");
		if (parent === current) break;
		current = parent;
	}
	return startPath;
}

async function evaluate(factsPath: string, modelDir: string) {
	const repoRoot = findRepoRoot(process.cwd());
	const absoluteFactsPath = resolve(repoRoot, factsPath);

	if (!existsSync(absoluteFactsPath)) {
		console.error(`${RED}Error:${RESET} Facts file not found: ${factsPath}`);
		process.exit(2);
	}

	console.log(`${BOLD}Evaluating${RESET} ${CYAN}${factsPath}${RESET} with model ${CYAN}${modelDir}${RESET}...\n`);

	const facts = await loadYamlFile<Facts>(absoluteFactsPath);
	const paths = modelPaths(repoRoot, modelDir);
	const version = await getModelVersion(repoRoot, modelDir);

	console.log(`${BOLD}Model version:${RESET} ${version}\n`);

	// Check for version mismatch
	const pinnedVersion = facts.model_version;
	if (typeof pinnedVersion === "string" && pinnedVersion.trim() && pinnedVersion.trim() !== version) {
		console.error(
			`${YELLOW}WARN:${RESET} facts model_version '${pinnedVersion.trim()}' does not match loaded model '${version}'\n`,
		);
	}

	const result = await evaluateFacts(facts, paths);

	printActivatedDomains(result.activated_domains, paths.questionsDir);
	printRequiredQuestions(result.required_questions);
	printDerivedControls(result.derived_controls);
}

function printActivatedDomains(domains: string[], questionsDir: string) {
	console.log(`${BOLD}Activated domains:${RESET}`);
	if (domains.length === 0) {
		console.log(`${DIM}  - (none)${RESET}`);
	} else {
		for (const domain of domains) {
			const filePath = `${questionsDir}/${domain}.questions.yaml`;
			console.log(`  ${GREEN}•${RESET} ${domain} ${DIM}(${filePath})${RESET}`);
		}
	}
	console.log();
}

function printRequiredQuestions(questions: Array<{ id: string; answered: boolean }>) {
	const missing = questions.filter((q) => !q.answered);
	const answered = questions.length - missing.length;

	console.log(`${BOLD}Required questions:${RESET}`);
	console.log(`  Total: ${questions.length}`);
	console.log(`  ${GREEN}Answered: ${answered}${RESET}`);
	console.log(`  ${missing.length > 0 ? YELLOW : GREEN}Missing: ${missing.length}${RESET}`);

	if (missing.length > 0) {
		console.log(`\n  ${BOLD}Missing answers:${RESET}`);
		// Group by domain
		const byDomain = new Map<string, string[]>();
		for (const q of missing) {
			const [domain, ...rest] = q.id.split(".");
			const qid = rest.join(".");
			if (!byDomain.has(domain)) byDomain.set(domain, []);
			byDomain.get(domain)!.push(qid);
		}

		for (const [domain, ids] of Array.from(byDomain.entries()).sort()) {
			console.log(`    ${BOLD}${domain}:${RESET}`);
			for (const id of ids) {
				console.log(`      ${YELLOW}○${RESET} ${id}`);
			}
		}
	}
	console.log();
}

function printDerivedControls(
	controls: Array<{
		id: string;
		title: string;
		scope: string;
		enforcement_intent: string;
		activation_phase: string;
		evidence_type: string[];
		because: Record<string, unknown>[];
		references?: Array<{ type: string; ref: string; description?: string }>;
	}>,
) {
	console.log(`${BOLD}Derived controls:${RESET}`);
	if (controls.length === 0) {
		console.log(`${DIM}  - (none)${RESET}\n`);
		return;
	}

	console.log(`  ${BOLD}Total: ${controls.length}${RESET}\n`);

	for (const c of controls) {
		console.log(`  ${CYAN}${BOLD}${c.id}${RESET}: ${c.title}`);
		console.log(`    ${DIM}scope:${RESET} ${c.scope}`);
		console.log(`    ${DIM}enforcement:${RESET} ${c.enforcement_intent}`);
		console.log(`    ${DIM}phase:${RESET} ${c.activation_phase}`);

		if (c.evidence_type.length > 0) {
			console.log(`    ${DIM}evidence:${RESET} ${c.evidence_type.join(", ")}`);
		}

		if (c.references && c.references.length > 0) {
			console.log(`    ${DIM}references:${RESET}`);
			for (const ref of c.references) {
				const desc = ref.description ? ` - ${ref.description}` : "";
				console.log(`      ${BLUE}${ref.type}${RESET}: ${ref.ref}${desc}`);
			}
		}

		console.log(`    ${DIM}because:${RESET}`);
		for (const why of c.because) {
			const condStr = Object.entries(why)
				.map(([k, v]) => `${k}=${JSON.stringify(v)}`)
				.join(", ");
			console.log(`      ${GREEN}•${RESET} when: { ${condStr} }`);
		}
		console.log();
	}
}

async function diff(factsPath: string, oldModelDir: string, newModelDir: string) {
	const repoRoot = findRepoRoot(process.cwd());
	const absoluteFactsPath = resolve(repoRoot, factsPath);

	if (!existsSync(absoluteFactsPath)) {
		console.error(`${RED}Error:${RESET} Facts file not found: ${factsPath}`);
		process.exit(2);
	}

	const facts = await loadYamlFile<Facts>(absoluteFactsPath);

	const oldPaths = modelPaths(repoRoot, oldModelDir);
	const newPaths = modelPaths(repoRoot, newModelDir);

	const oldVersion = await getModelVersion(repoRoot, oldModelDir);
	const newVersion = await getModelVersion(repoRoot, newModelDir);

	console.log(`${BOLD}Diff:${RESET} ${CYAN}${oldModelDir}${RESET} (${oldVersion}) → ${CYAN}${newModelDir}${RESET} (${newVersion})\n`);

	const oldResult = await evaluateFacts(facts, oldPaths);
	const newResult = await evaluateFacts(facts, newPaths);

	printControlsDiff(oldResult, newResult);
	printQuestionsDiff(oldResult, newResult);
	printDomainsDiff(oldResult, newResult);
}

function printControlsDiff(oldResult: EvaluateResult, newResult: EvaluateResult) {
	const oldControls = new Set(oldResult.derived_controls.map((c) => c.id));
	const newControls = new Set(newResult.derived_controls.map((c) => c.id));

	const added = [...newControls].filter((id) => !oldControls.has(id)).sort();
	const removed = [...oldControls].filter((id) => !newControls.has(id)).sort();

	console.log(`${BOLD}Controls:${RESET}`);
	if (added.length === 0 && removed.length === 0) {
		console.log(`${DIM}  - (no changes)${RESET}`);
	} else {
		if (added.length > 0) {
			console.log(`  ${GREEN}+ added (${added.length}):${RESET}`);
			for (const id of added) {
				const control = newResult.derived_controls.find((c) => c.id === id);
				console.log(`      ${id} - ${control?.title ?? "(unknown)"}`);
			}
		}
		if (removed.length > 0) {
			console.log(`  ${RED}- removed (${removed.length}):${RESET}`);
			for (const id of removed) {
				const control = oldResult.derived_controls.find((c) => c.id === id);
				console.log(`      ${id} - ${control?.title ?? "(unknown)"}`);
			}
		}
	}
	console.log();
}

function printQuestionsDiff(oldResult: EvaluateResult, newResult: EvaluateResult) {
	const oldMissing = new Set(oldResult.required_questions.filter((q) => !q.answered).map((q) => q.id));
	const newMissing = new Set(newResult.required_questions.filter((q) => !q.answered).map((q) => q.id));

	const newlyMissing = [...newMissing].filter((id) => !oldMissing.has(id)).sort();
	const noLongerMissing = [...oldMissing].filter((id) => !newMissing.has(id)).sort();

	console.log(`${BOLD}Required questions (missing):${RESET}`);
	if (newlyMissing.length === 0 && noLongerMissing.length === 0) {
		console.log(`${DIM}  - (no changes)${RESET}`);
	} else {
		if (newlyMissing.length > 0) {
			console.log(`  ${YELLOW}+ newly missing (${newlyMissing.length}):${RESET} ${newlyMissing.join(", ")}`);
		}
		if (noLongerMissing.length > 0) {
			console.log(`  ${GREEN}- no longer missing (${noLongerMissing.length}):${RESET} ${noLongerMissing.join(", ")}`);
		}
	}
	console.log();
}

function printDomainsDiff(oldResult: EvaluateResult, newResult: EvaluateResult) {
	const oldDomains = oldResult.activated_domains.join(", ");
	const newDomains = newResult.activated_domains.join(", ");

	console.log(`${BOLD}Activated domains:${RESET}`);
	if (oldDomains === newDomains) {
		console.log(`${DIM}  - (unchanged)${RESET} ${oldDomains}`);
	} else {
		console.log(`  ${RED}old:${RESET} ${oldDomains}`);
		console.log(`  ${GREEN}new:${RESET} ${newDomains}`);
	}
	console.log();
}

async function main() {
	const { values, positionals } = parseArgs({
		args: Bun.argv.slice(2),
		options: {
			"model-dir": { type: "string", default: "model" },
			old: { type: "string" },
			new: { type: "string" },
			help: { type: "boolean", short: "h" },
		},
		allowPositionals: true,
	});

	if (values.help || positionals.length === 0) {
		console.log(`${BOLD}Risk Assistant CLI${RESET}

${BOLD}Usage:${RESET}
  bun tools/riskctl.ts evaluate <facts.yaml> [--model-dir model]
  bun tools/riskctl.ts diff <facts.yaml> --old model-v1 --new model-v2

${BOLD}Commands:${RESET}
  ${CYAN}evaluate${RESET}  Evaluate a facts file against a model
  ${CYAN}diff${RESET}      Compare outcomes between two model versions

${BOLD}Options:${RESET}
  --model-dir <dir>  Path to model directory (default: model)
  --old <dir>        Old model directory (for diff)
  --new <dir>        New model directory (for diff)
  -h, --help         Show this help message
`);
		process.exit(values.help ? 0 : 1);
	}

	const command = positionals[0];

	try {
		if (command === "evaluate") {
			const factsPath = positionals[1];
			if (!factsPath) {
				console.error(`${RED}Error:${RESET} Missing facts file argument\n`);
				console.error("Usage: bun tools/riskctl.ts evaluate <facts.yaml> [--model-dir model]");
				process.exit(1);
			}
			await evaluate(factsPath, values["model-dir"]!);
		} else if (command === "diff") {
			const factsPath = positionals[1];
			if (!factsPath || !values.old || !values.new) {
				console.error(`${RED}Error:${RESET} Missing required arguments\n`);
				console.error("Usage: bun tools/riskctl.ts diff <facts.yaml> --old model-v1 --new model-v2");
				process.exit(1);
			}
			await diff(factsPath, values.old, values.new);
		} else {
			console.error(`${RED}Error:${RESET} Unknown command '${command}'\n`);
			console.error("Available commands: evaluate, diff");
			console.error("Run with --help for usage information");
			process.exit(1);
		}
	} catch (error) {
		console.error(`${RED}Error:${RESET}`, error instanceof Error ? error.message : error);
		if (error instanceof Error && error.stack) {
			console.error(`\n${DIM}${error.stack}${RESET}`);
		}
		process.exit(1);
	}
}

main();
